src/inventory_ui.cpp

```diff
@@ -209,99 +345,157 @@ void inventory_column::prepare_paging( size_t new_entries_per_page )
             ? inventory_entry() // the last item on the page must not be a category
             : inventory_entry( current_category ); // the first item on the page must be a category
         entries.insert( entries.begin() + i, insertion );
+        expand_to_fit( insertion );
     }
+
+    paging_is_valid = true;
 }
 
-std::string inventory_column::get_entry_text( const inventory_entry &entry ) const
+void inventory_column::remove_entry( const inventory_entry &entry )
 {
-    std::ostringstream res;
-
-    if( entry.is_item() ) {
-        if( get_option<bool>( "ITEM_SYMBOLS" ) ) {
-            res << entry.get_item().symbol() << ' ';
-        }
-
-        if( multiselect ) {
-            if( entry.chosen_count == 0 ) {
-                res << "<color_c_dkgray>-";
-            } else if( entry.chosen_count >= entry.get_available_count() ) {
-                res << "<color_c_ltgreen>+";
-            } else {
-                res << "<color_c_ltgreen>#";
-            }
-            res << " </color>";
-        }
-
-        const size_t count = entry.get_stack_size();
-        if( count > 1 ) {
-            res << count << ' ';
-        }
-        res << entry.get_item().display_name( count );
-    } else if( entry.get_category_ptr() != nullptr ) {
-        res << entry.get_category_ptr()->name;
+    const auto iter = std::find( entries.begin(), entries.end(), entry );
+    if( iter == entries.end() ) {
+        debugmsg( "Tried to remove a non-existing entry." );
+        return;
     }
-    return res.str();
+    entries.erase( iter );
+    paging_is_valid = false;
 }
 
-nc_color inventory_column::get_entry_color( const inventory_entry &entry ) const
+void inventory_column::clear()
 {
-    if( entry.is_item() ) {
-        return ( active && is_selected( entry ) )
-            ? h_white
-            : ( custom_colors && entry.custom_color != c_unset )
-                ? entry.custom_color
-                : entry.get_item().color_in_inventory();
-    } else {
-        return c_magenta;
-    }
+    entries.clear();
+    paging_is_valid = false;
+    prepare_paging();
 }
 
 size_t inventory_column::get_entry_indent( const inventory_entry &entry ) const {
-    return entry.is_item() ? 2 : 0;
+    if( !entry.location ) {
+        return 0;
+    }
+    size_t res = 2;
+    if( get_option<bool>("ITEM_SYMBOLS" ) ) {
+        res += 2;
+    }
+    if( allows_selecting() && multiselect ) {
+        res += 2;
+    }
+    return res;
 }
 
-size_t inventory_column::get_entry_width( const inventory_entry &entry ) const {
-    return get_entry_indent( entry ) + utf8_width( get_entry_text( entry ), true );
+long inventory_column::reassign_custom_invlets( const player &p, long min_invlet, long max_invlet )
+{
+    long cur_invlet = min_invlet;
+    for( auto &entry : entries ) {
+        if( entry.location && !p.has_item( *entry.location ) ) {
```

Comment:
Checking validity via `entry.location` includes a check that the item is still in player posession/on map tile/ in vehicle. Given that does this code do what you intend?